caas/kubernetes/provider/exec/exec.go

```diff
@@ -0,0 +1,248 @@
+func New(namesapce string, clientset kubernetes.Interface, config *rest.Config) Executor {
+	return new(
+		namesapce,
+		clientset,
+		config,
+		remotecommand.NewSPDYExecutor,
+		func() (io.Reader, io.WriteCloser) { return io.Pipe() },
+	)
+}
+
+func new(
+	namesapce string,
+	clientset kubernetes.Interface,
+	config *rest.Config,
+	remoteCMDNewer func(config *rest.Config, method string, url *url.URL) (remotecommand.Executor, error),
+	pipGetter func() (io.Reader, io.WriteCloser),
+) Executor {
+	return &client{
+		namesapce: namesapce,
+		clientset: clientset,
+		remoteCmdExecutorGetter: func(method string, url *url.URL) (remotecommand.Executor, error) {
+			return remoteCMDNewer(config, method, url)
+		},
+		podGetter: clientset.CoreV1().Pods(namesapce),
+		pipGetter: pipGetter,
+	}
+}
+
+// ExecParams holds all the necessary parameters for Exec.
+type ExecParams struct {
+	Commands      []string
+	Env           []string
+	PodName       string
+	ContainerName string
+	WorkingDir    string
+
+	Stdin  io.Reader
+	Stdout io.Writer
+	Stderr io.Writer
+}
+
+func (ep *ExecParams) validate(podGetter typedcorev1.PodInterface) (err error) {
+	if len(ep.Commands) == 0 {
+		return errors.NotValidf("empty commands")
+	}
+
+	if ep.PodName, ep.ContainerName, err = getValidatedPodContainer(
+		podGetter, ep.PodName, ep.ContainerName,
+	); err != nil {
+		return errors.Trace(err)
+	}
+	return nil
+}
+
+// Exec runs commands on a pod in the cluster.
+func (c client) Exec(params ExecParams, cancel <-chan struct{}) error {
+	if err := params.validate(c.podGetter); err != nil {
+		return errors.Trace(err)
+	}
+	return errors.Trace(c.exec(params, cancel))
+}
+
+func processEnv(env []string) string {
+	out := ""
+	for _, v := range env {
+		out += fmt.Sprintf("export %s; ", v)
+	}
+	return out
+}
+
+func (c client) exec(opts ExecParams, cancel <-chan struct{}) error {
+	cmd := ""
+	if opts.WorkingDir != "" {
+		cmd += fmt.Sprintf("cd %s; ", opts.WorkingDir)
+	}
+	if len(opts.Env) > 0 {
+		cmd += processEnv(opts.Env)
+	}
+	cmd += fmt.Sprintf("%s; ", strings.Join(opts.Commands, " "))
+	cmdArgs := []string{"sh", "-c", cmd}
+
+	req := c.clientset.CoreV1().RESTClient().Post().
+		Resource("pods").
+		Name(opts.PodName).
+		Namespace(c.namesapce).
+		SubResource("exec").
+		Param("container", opts.ContainerName).
+		VersionedParams(&core.PodExecOptions{
+			Container: opts.ContainerName,
+			Command:   cmdArgs,
+			Stdin:     opts.Stdin != nil,
+			Stdout:    opts.Stdout != nil,
+			Stderr:    opts.Stderr != nil,
+			TTY:       false,
+		}, scheme.ParameterCodec)
+
+	executor, err := c.remoteCmdExecutorGetter("POST", req.URL())
+	if err != nil {
+		return errors.Trace(err)
+	}
+
+	errChan := make(chan error, 1)
+	go func() {
+		errChan <- executor.Stream(remotecommand.StreamOptions{
+			Stdin:  opts.Stdin,
+			Stdout: opts.Stdout,
+			Stderr: opts.Stderr,
+			Tty:    false,
+		})
+	}()
+	select {
+	case err := <-errChan:
+		return errors.Trace(err)
+	case <-cancel:
```

Comment:
req should have a context that is cancelled.
See k8s.io/client-go/rest/request.go:416

Reply:
ah, we just use request to genenrate the URL.
what we use is `SPDYExecutor` created by`NewSPDYExecutor`.